src/server/tiering/op_manager.cc

```diff
@@ -0,0 +1,122 @@
+// Copyright 2024, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "server/tiering/op_manager.h"
+
+#include "base/logging.h"
+#include "core/overloaded.h"
+#include "io/io.h"
+#include "server/tiering/common.h"
+
+namespace dfly::tiering {
+
+namespace {
+
+OpManager::OwnedEntryId ToOwned(OpManager::EntryId id) {
+  Overloaded convert{[](unsigned i) -> OpManager::OwnedEntryId { return i; },
+                     [](std::string_view s) -> OpManager::OwnedEntryId { return std::string{s}; }};
+  return std::visit(convert, id);
+}
+
+OpManager::EntryId Borrowed(const OpManager::OwnedEntryId& id) {
+  return std::visit([](const auto& v) -> OpManager::EntryId { return v; }, id);
+}
+
+}  // namespace
+
+std::error_code OpManager::Open(std::string_view file) {
+  return storage_.Open(file);
+}
+
+void OpManager::Close() {
+  storage_.Close();
+}
+
+util::fb2::Future<std::string> OpManager::Read(EntryId id, DiskSegment segment) {
+  return PrepareRead(segment).ForId(id, segment).futures.emplace_back().get_future();
+}
+
+void OpManager::Delete(EntryId id) {
+  // If the item isn't offloaded, it has io pending, so cancel it
+  DCHECK(pending_stash_ver_.count(ToOwned(id)));
+  ++pending_stash_ver_[ToOwned(id)];
+}
+
+void OpManager::Delete(DiskSegment segment) {
+  if (auto it = pending_reads_.find(segment.offset); it != pending_reads_.end()) {
+    // If a read is pending, it will be deleted once the read finished
+    DCHECK_EQ(it->second.segment.length, segment.length);
+    it->second.delete_requested = true;
+  } else {
+    // Otherwise, delete it immediately
+    storage_.MarkAsFree(segment);
+  }
+}
+
+std::error_code OpManager::Stash(EntryId id_ref, std::string_view value) {
+  auto id = ToOwned(id_ref);
+  unsigned version = ++pending_stash_ver_[id];
+
+  std::shared_ptr<char[]> buf(new char[value.length()]);
```

Comment:
what's the reason for copying here? 
how many buf copied do we when we stash a PrimeValue?

Reply:
The reason for copying is that we can't guarantee that the value will be valid while the write is pending - outdated design of DiskStorage. I didn't yet update DiskStorage's interface, it should make a copy of the value (copy to internal buffer or to registered buffer). Will do next time

Reply:
Will expand on that with registered buffers, we might actually want to call GetString on CO into the buffer directly

Reply:
I suggest to adjust the api via DiskStorage without relying on registered buffers API. can be in the next PR.